skimage/segmentation/_thresholding_global.py

```diff
@@ -0,0 +1,984 @@
+         If ``image`` contains less grayscale value then the desired
+         number of classes.
+
+    Notes
+    -----
+    This implementation relies on a Cython function whose complexity
+    is :math:`O\left(\frac{Ch^{C-1}}{(C-1)!}\right)`, where :math:`h`
+    is the number of histogram bins and :math:`C` is the number of
+    classes desired.
+
+    If no hist is given, this function will make use of
+    `skimage.exposure.histogram`, which behaves differently than
+    `np.histogram`. While both allowed, use the former for consistent
+    behaviour.
+
+    The input image must be grayscale.
+
+    References
+    ----------
+    .. [1] Liao, P-S., Chen, T-S. and Chung, P-C., "A fast algorithm for
+           multilevel thresholding", Journal of Information Science and
+           Engineering 17 (5): 713-727, 2001. Available at:
+           <https://ftp.iis.sinica.edu.tw/JISE/2001/200109_01.pdf>
+           :DOI:`10.6688/JISE.2001.17.5.1`
+    .. [2] Tosa, Y., "Multi-Otsu Threshold", a java plugin for ImageJ.
+           Available at:
+           <http://imagej.net/plugins/download/Multi_OtsuThreshold.java>
+
+    Examples
+    --------
+    >>> from skimage.color import label2rgb
+    >>> from skimage import data
+    >>> image = data.camera()
+    >>> thresholds = threshold_multiotsu(image)
+    >>> regions = np.digitize(image, bins=thresholds)
+    >>> regions_colorized = label2rgb(regions)
+    """
+    if image is not None and image.ndim > 2 and image.shape[-1] in (3, 4):
+        warn(
+            f'threshold_multiotsu is expected to work correctly only for '
+            f'grayscale images; image shape {image.shape} looks like '
+            f'that of an RGB image.'
+        )
+
+    # calculating the histogram and the probability of each gray level.
+    prob, bin_centers = _validate_image_histogram(image, hist, nbins, normalize=True)
+    prob = prob.astype('float32', copy=False)
+
+    nvalues = np.count_nonzero(prob)
+    if nvalues < classes:
+        msg = (
+            f'After discretization into bins, the input image has '
+            f'only {nvalues} different values. It cannot be thresholded '
+            f'in {classes} classes. If there are more unique values '
+            f'before discretization, try increasing the number of bins '
+            f'(`nbins`).'
+        )
+        raise ValueError(msg)
+    elif nvalues == classes:
+        thresh_idx = np.flatnonzero(prob)[:-1]
+    else:
+        # Get threshold indices
+        try:
+            thresh_idx = _get_multiotsu_thresh_indices_lut(prob, classes - 1)
+        except MemoryError:
+            # Don't use LUT if the number of bins is too large (if the
+            # image is uint16 for example): in this case, the
+            # allocated memory is too large.
+            thresh_idx = _get_multiotsu_thresh_indices(prob, classes - 1)
+
+    thresh = bin_centers[thresh_idx]
+
+    return thresh
+
+
+def _cross_entropy(image, *, threshold, bins=None):
```

Comment:
Note the change to the default value of `bin=` in https://github.com/scikit-image/scikit-image/pull/7781/commits/bfe7908cdca322a6d037fffdcdaa75eaa2006ac6.